guidance/_grammar.py

```diff
@@ -494,37 +447,31 @@ def commit_point(value, hidden=False):
     raise NotImplementedError("commit_point is not implemented (may remove in the future)")
 
 
-def as_regular_grammar(value):
-    # TODO: assert that value is not empty since we don't yet support that
-    if isinstance(value, str):
-        value = string(value)
-    # check if it serializes
-    _ignore = LLSerializer().regex(value)
-    return RegularGrammar(value)
-
-
 class Placeholder(GrammarFunction):
     def __init__(self):
-        pass
+        super().__init__(capture_name=None)
 
 
 class Join(GrammarFunction):
     __slots__ = (
         "values",
         "name",
-        "capture_name",
         "max_tokens",
     )
 
     def __init__(
-        self, values, name: Union[str, None] = None, max_tokens=100000000
+        self,
+        values: Sequence[ComposableGrammar],
+        name: Union[str, None] = None,
+        max_tokens=100000000,
     ) -> None:
-        values = [
-            string(v) if isinstance(v, (str, bytes)) else v for v in values
-        ]  # wrap raw strings
-        self.values = [v for v in values if not isinstance(v, Null)]
+        super().__init__(capture_name=None)
+        # wrap raw strings
+        converted_values = [string(v) if isinstance(v, (str, bytes)) else v for v in values]
```

Comment:
TBH, I'd prefer to have `Select` and `Join` be strict in their accepted types, and have all the type conversions done by `select()` and `join()`.